### PR TITLE
Pulse

### LINKED ISSUE #260: Enhance the Pulse-Level Simulation
We want to continue working to build up our pulse-level simulation. You can reference an example from [QuTip](https://qutip.org/docs/4.6/guide/qip/qip-processor.html) and see our current implementation in pulse.

### PR DESCRIPTION
**Pull request by AbdullahKazi500**
QuantumPulseDemo Class:

Inherits from tq.QuantumModule, a PyTorch module for quantum operations.
Sets up a quantum encoder and a parameterized quantum pulse.
apply_pulse Method:

Applies the parameterized quantum pulse to a quantum device.
Iterates over pulse parameters and applies them to each wire.
main Function:

Initializes the model and optimizer.
Defines the target unitary matrix.
Enters a training loop over epochs.
Computes loss based on the difference between the current and target unitary matrices.
Performs backpropagation and optimization.
Prints progress and current pulse shape/unitary matrix.
Feature Request:

This code showcases how to train a quantum pulse to approximate a target unitary operation using gradient-based optimization.
It's a feature request for optimizing quantum pulses using PyTorch, which can be useful for various quantum tasks such as quantum gate synthesis, quantum state preparation, and quantum error correction.

QuantumPulseDemo Class:

Inherits from tq.QuantumModule, facilitating quantum operations within a PyTorch framework.
Utilizes a quantum encoder (tq.GeneralEncoder) to encode input data into quantum states.
Defines a parameterized quantum pulse (tq.pulse.QuantumPulseDirect) with a specified number of steps and Hamiltonian.
apply_pulse Method:

Applies the parameterized quantum pulse to a quantum device.
Iterates over pulse parameters and applies them to each wire using tqf.rx.
main Function:

Sets up input data and a target unitary matrix.
Initializes the model and optimizer.
Enters a training loop over epochs.
Performs a forward pass through the model to compute the loss.
Computes the loss based on the difference between the current and target unitary matrices.
Performs backpropagation and optimization using the Adam optimizer.
Prints progress and current pulse shape/unitary matrix every 100 epochs.
Feature Request:

This code showcases how to train a quantum pulse to approximate a target unitary operation using gradient-based optimization in a PyTorch environment.
It demonstrates a practical method for optimizing quantum pulses, which is a crucial feature for various quantum computing tasks such as quantum gate synthesis, quantum state preparation, and quantum error correction.

The code consists of three main components:

QuantumPulseDemo Class: This class defines a quantum module that demonstrates the application of a parameterized quantum pulse to a quantum device. It utilizes the TorchQuantum library to define the quantum encoder and parameterized quantum pulse. The module is trained to optimize the parameters of the quantum pulse to approximate a target unitary operation.

OM_EOM_Simulation Class: This class simulates a sequence of optical pulses, optionally with electro-optic modulation (EOM). It generates a sequence of pulses and delays, which can be used to model various optical communication scenarios.

QuantumPulseDemoRunner Class: This class serves as a runner for training the QuantumPulseDemo model and simulating the OM_EOM_Simulation. It initializes the model, optimizer, target unitary, and simulator based on user-provided parameters. It provides methods to train the quantum pulse demonstration model and run the optical modulation simulation.
fixes #260

### CODE REVIEW

**Review by GenericP3rson — CHANGES_REQUESTED** [review]
Hello! Could you update your notebook to instead be a class in the pulse folder? Thanks!

File: torchQuantumpulse.ipynb
Comment (GenericP3rson) [inline torchQuantumpulse.ipynb:1]: This can be removed

File: torchquantum/pulse/quantum_pulse_simulation.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    This module defines a parameterized quantum pulse and applies it to a quantum device.
+    """
+
+    def __init__(self):
+        """
+        Initializes the QuantumPulseDemo module.
+        """
+        super().__init__()
+        self.n_wires = 2
```
Comment (GenericP3rson) [inline torchquantum/pulse/quantum_pulse_simulation.py:19]: I would recommend parametrising the function. If you're doing an example, it should be in the examples folder. I'll wait for @Hanrui-Wang to review for his suggestions.
